src/os/bluestore/BlueStore.cc

```diff
@@ -8877,84 +8877,78 @@ string BlueStore::get_device_path(unsigned id)
   return res;
 }
 
-int BlueStore::_set_bdev_label_size(const string& path, uint64_t size)
-{
-  bluestore_bdev_label_t label;
-  int r = _read_bdev_label(cct, bdev, path, &label);
-  if (r < 0) {
-    derr << "unable to read label for " << path << ": "
-          << cpp_strerror(r) << dendl;
-  } else {
-    label.size = size;
-    r = _write_bdev_label(cct, bdev, path, label);
-    if (r < 0) {
-      derr << "unable to write label for " << path << ": "
-            << cpp_strerror(r) << dendl;
-    }
-  }
-  return r;
-}
-
 int BlueStore::expand_devices(ostream& out)
 {
   int r = _open_db_and_around(true);
   ceph_assert(r == 0);
   bluefs->dump_block_extents(out);
   out << "Expanding DB/WAL..." << std::endl;
-  for (auto devid : { BlueFS::BDEV_WAL, BlueFS::BDEV_DB}) {
-    if (devid == bluefs_layout.shared_bdev ) {
+  uint64_t old_shared_size = 0;
+  uint64_t new_shared_size = 0;
+  for (auto devid : { BlueFS::BDEV_WAL, BlueFS::BDEV_DB, BlueFS::BDEV_SLOW}) {
+    bool shared_dev = devid == bluefs_layout.shared_bdev;
+    auto bdev0 = shared_dev ? bdev : bluefs->get_block_device(devid);
+    if (!bdev0 || new_shared_size) {
       continue;
     }
-    uint64_t size = bluefs->get_block_device_size(devid);
-    if (size == 0) {
-      // no bdev
+    bluestore_bdev_label_t label;
+    bool label_supported = bdev0->supported_bdev_label();
+    uint64_t new_size = bdev0->get_size();
+    int r = label_supported ? _read_bdev_label(cct, bdev0, path, &label) : 0;
+    if (r < 0) {
+      derr << devid
+           << " : unable to read label for " << path << ": "
+           << cpp_strerror(r) << dendl;
+      continue;
+    }
+    uint64_t size = shared_dev ? fm->get_size() : label.size;
+    if (p2align(new_size, min_alloc_size) <= size) {
+      // no need to expand
+      out << devid
+	  <<" : nothing to do, skipped"
+	  << std::endl;
       continue;
     }
-
-    out << devid
-	<<" : expanding " << " to 0x" << size << std::dec << std::endl;
     string p = get_device_path(devid);
     const char* path = p.c_str();
     if (path == nullptr) {
       derr << devid
-	    <<": can't find device path " << dendl;
+           <<": can't find device path " << dendl;
       continue;
     }
-    if (bluefs->bdev_support_label(devid)) {
-      if (_set_bdev_label_size(p, size) >= 0) {
-        out << devid
-          << " : size label updated to " << size
+    out << devid
+	<<" : Expanding to 0x" << std::hex << new_size
+	<< std::dec << "(" << byte_u_t(new_size) << ")"
+	<< std::endl;
+    label.size = new_size;
+    r = label_supported ? _write_bdev_label(cct, bdev0, path, label) : 0;
```

Comment:
There is now a mess that I introduced.
When operating on BDEV_SLOW it should incorporate multi bdev label, like in `BlueStore::write_meta()`
```
_read_multi_bdev_label(...,&bdev_label_valid_locations, ...);
....
int r = _write_bdev_label(..., bdev_label_valid_locations);
```

Reply:
More over, if new locations became available for label clones we should modify `bdev_label_valid_locations`.

Reply:
In fact, maybe above should be addressed by me in a some tracker.